### ISSUE TITLE
Benchmark not reproducible

### ISSUE THREAD

**Issue by jzhou316** [pyg-team/pytorch_geometric#859]
## Benchmark for graph classification (e.g. with SAGPool) not reproducible

Hi I was running the benchmarks for graph classification in here https://github.com/rusty1s/pytorch_geometric/tree/master/benchmark/kernel

However, I found that the runs of the same model, same data split, have quite different results so there is no way to replicate a model, even if I set the same random seeds and make sure the two runs are identical exactly. For example, I set the model to be SAGPool, with 4 layers, 64 hidden, and the random seeds to be 0, the training and test accuracy is quite different for multiple runs. 

I checked in detail how that happened, and found that the GraphConv layers output slightly different values at every run, together with softmax, topk, the final difference could be relatively huge. Any idea of the reason and how to make the benchmark experiments reproducible?

**Comment by rusty1s** [comment]
You are right that the aggregation scheme is currently non-deterministic. There are a lot of open issues concerning this and I am working on it by providing support for `segment_*` operations.

However, IMO I do not find this troublesome at all. In the end, there is no clear ordering of your neighborhood that one should rely on when aggregating, and by using non-deterministic operations you should get a much more robust model w.r.t. to permutations in your input graph. Furthermore, this all equals out when reporting the average results across a number of experiments, which one should *always* do.

**Comment by tchaton** [comment]
Hey @rusty1s, could I contribute to the segment implementation project. There is also the HAG on my repo in process.

**Comment by rusty1s** [comment]
You surely can :) I will provide a first implementation for `segment_add` in the `torch_scatter` repo next week, which you could enhance or benchmark against your own solution. Would that work for you?

I know about your HAG project and I am looking forward to it.

**Comment by tchaton** [comment]
Hey @rusty1s, Sounds great ! Do you have a slack channel for coordinating the work to be done ? If not, I have just created one where the work on pytorch geometric and related project could be talked about. I sent you an invite.

**Comment by jzhou316** [comment]
Hi @rusty1s thanks for the clarification! Can you elaborate a bit more about the high-levels of why the non-deterministic behavior for things like scatter_add? I may be dumb since I don't know much about the low-level implementation. In principle even if the ordering of a neighborhood is different, the aggregation results should always be the same right? Otherwise this impact on the model output no longer makes the process ordering-invariant which is what we want, and it will be less robust on the testing case, since we could possibly have the same graph, represented in different ways, but with different classification results.

**Comment by rusty1s** [comment]
Sure! Sorry for not being clear the first time. Scatter operations use atomic operations under the hood, and hence the ordering of summation is not always the same. However, floating point arithmetic is actually non-associative. [Here](https://www.youtube.com/watch?v=7xCkXhwU3tI&list=PLAwxTw4SYaPnFKojVQrmyOGFCqHTxfdv2&index=104) is a good video describing this phenomenon.

**Comment by jzhou316** [comment]
Thanks! This solves my confusion.

**Comment by MaryamHamad** [comment]
How did you solve this issue?
